File: app/src/main/java/org/techtown/retrofit2/MovieAdapter.java
package org.techtown.retrofit2;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieViewHolder> {
    ArrayList<Movie> items = new ArrayList();

    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View itemView = inflater.inflate(R.layout.movie_item, parent, false);

        return new MovieViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        Movie item = items.get(position);
        holder.setItem(item);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void addItem(Movie item) {
        items.add(item);
    }

    public Movie getItem(int position) {
        return items.get(position);
    }

    public void addItems(List<Movie> items) {
        this.items.addAll(items);
        notifyDataSetChanged();
        //TODO 추가된 영역만 갱신
        //notifyItemRangeInserted();
    }

    public void removeItem(Movie item) {
        //TODO 아이템 삭제
    }

    public ArrayList<Movie> getItems() {
        return items;
    }

    public void clearItems() {
        this.items.clear();
        notifyDataSetChanged();
    }

    static class MovieViewHolder extends RecyclerView.ViewHolder {
        TextView textView;
        TextView textView2;
        //TODO 삭제 뷰 추가

        public MovieViewHolder(@NonNull View itemView) {
            super(itemView);

            textView = itemView.findViewById(R.id.textView_adapter_movie);
            textView2 = itemView.findViewById(R.id.textView2_adapter_movie);
        }

        public void setItem(Movie item) {
            textView.setText(item.getMovieNm());
            textView2.setText(item.getGenreAlt() + " 명");
            //TODO 삭제 뷰 onClickListener 붙여줌 , adapter 에 있는 items 를 viewholder 에서 알수가 없다. 어떻게 해결할까!?

        }
    }

}

File: app/src/main/java/org/techtown/retrofit2/MovieAdapter.kt
package org.techtown.retrofit2

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class MovieAdapter(private val movieItemClickListener: MovieClickListener) :
    RecyclerView.Adapter<MovieViewHolder>() {
    var items: ArrayList<Movie> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val itemView = inflater.inflate(R.layout.movie_item, parent, false)
        return MovieViewHolder(itemView, movieItemClickListener)
    }

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        val item = items[position]
        holder.setItem(item)
    }

    override fun getItemCount(): Int {
        return items.size
    }

    fun addItem(item: Movie) {
        items.add(item)
    }

    fun getItem(position: Int): Movie {
        return items[position]
    }

    fun addItems(items: List<Movie>) {
        val positionStart: Int = this.items.size + 1
        this.items.addAll(items)
        notifyItemRangeInserted(positionStart, items.size)
    }

    fun removeItem(item: Movie) {
        val positionstart = items.indexOf(item)
        removeItem(positionstart)
    }

    fun removeItem(position: Int) {
        items.removeAt(position)
        notifyItemRemoved(position)
        notifyItemRangeChanged(position, items.size - position - 1)
    }

    fun clearItems() {
        items.clear()
        notifyDataSetChanged()
    }

    fun favoriteItem(movie: Movie) {
        movie.isFavorite = movie.isFavorite.not()
        val position = items.indexOf(movie)
        notifyItemChanged(position)
    }


}

class MovieViewHolder(itemView: View, private val movieItemClickListener: MovieClickListener) :
    RecyclerView.ViewHolder(itemView) {
    var textView: TextView
    var textView2: TextView
    var button: Button
    var buttonFavorite: Button

    init {
        textView = itemView.findViewById(R.id.textView_adapter_movie)
        textView2 = itemView.findViewById(R.id.textView2_adapter_movie)
        button = itemView.findViewById(R.id.btn_remove)
        buttonFavorite = itemView.findViewById(R.id.btn_favorite)
    }

    fun setItem(item: Movie) {
        textView.text = item.movieNm
        textView2.text = item.genreAlt + " 명"
        button.setOnClickListener {
            movieItemClickListener.onClick(it, item)
        }
        buttonFavorite.setOnClickListener {
            movieItemClickListener.onClick(it, item)
        }

        if (item.isFavorite) {
            //favorite image resource
        } else {
            //not favorite image resource
        }
    }

}







